Symfony/CS/Fixer/PSR2/EofEndingFixer.php

```diff
@@ -23,15 +25,25 @@ class EofEndingFixer extends AbstractFixer
      */
     public function fix(\SplFileInfo $file, $content)
     {
-        // [Structure] A file must always end with a linefeed character
+        $tokens = Tokens::fromCode($content);
 
-        $content = rtrim($content);
+        $count = $tokens->count();
+        if (0 === $count) {
+            return;
+        }
+
+        $token = $tokens[$count - 1];
+        if ($token->isGivenKind(T_INLINE_HTML) || $token->isGivenKind(T_CLOSE_TAG)) {
+            return $content;
+        }
 
-        if ('' !== $content) {
-            return $content."\n";
+        if ($token->isGivenKind(T_WHITESPACE)) {
+            $token->setContent("\n");
+        } else {
+            $tokens->insertAt($count, new Token(array(T_WHITESPACE, "\n")));
         }
```

Comment:
What to do with a file containing only the open tag and a newline character? `<?php\n`
The newline is part of the `T_OPEN_TAG`. So this fixer would insert another newline character, but IMO it should not.